be/src/vec/exec/format/parquet/parquet_thrift_util.h

```diff
@@ -36,26 +36,62 @@ constexpr uint8_t PARQUET_VERSION_NUMBER[4] = {'P', 'A', 'R', '1'};
 
-static Status parse_thrift_footer(io::FileReaderSPtr file, FileMetaData** file_metadata,
-                                  size_t* meta_size, io::IOContext* io_ctx) {
+static Status read_thrift_footer(io::FileReaderSPtr file, std::vector<uint8_t>& footer,
+                                 size_t bytes_read, io::IOContext* io_ctx) {
     size_t file_size = file->size();
-    size_t bytes_read = std::min(file_size, INIT_META_SIZE);
-    std::vector<uint8_t> footer(bytes_read);
-    RETURN_IF_ERROR(file->read_at(file_size - bytes_read, Slice(footer.data(), bytes_read),
-                                  &bytes_read, io_ctx));
+    bytes_read = std::min(file_size, INIT_META_SIZE);
+    footer.reserve(bytes_read);
+    return file->read_at(file_size - bytes_read, Slice(footer.data(), bytes_read), &bytes_read,
+                         io_ctx);
+}
 
-    // validate magic
-    uint8_t* magic_ptr = footer.data() + bytes_read - 4;
-    if (bytes_read < PARQUET_FOOTER_SIZE ||
-        memcmp(magic_ptr, PARQUET_VERSION_NUMBER, sizeof(PARQUET_VERSION_NUMBER)) != 0) {
-        return Status::Corruption(
-                "Invalid magic number in parquet file, bytes read: {}, file size: {}, path: {}, "
-                "read magic: {}",
-                bytes_read, file_size, file->path().native(),
-                std::string((char*)magic_ptr, sizeof(PARQUET_VERSION_NUMBER)));
+static bool validate_magic_number(const std::vector<uint8_t>& footer, size_t bytes_read) {
+    if (footer.size() < PARQUET_FOOTER_SIZE) {
+        return false;
+    }
+    const uint8_t* magic_ptr = footer.data() + bytes_read - 4;
+    return memcmp(magic_ptr, PARQUET_VERSION_NUMBER, sizeof(PARQUET_VERSION_NUMBER)) == 0;
+}
+
+static Status parse_thrift_footer(io::FileReaderSPtr file, FileMetaData** file_metadata,
+                                  size_t* meta_size, io::IOContext* io_ctx) {
+    std::vector<uint8_t> footer;
+    size_t bytes_read = 0;
+    RETURN_IF_ERROR(read_thrift_footer(file, footer, bytes_read, io_ctx));
+    size_t before_file_size = file->size();
+    if (!validate_magic_number(footer, bytes_read)) {
+        // try to get file length from filesystem and validate magic again
+        Status st = file->update_size();
+        if (st.ok()) {
+            size_t file_size = file->size();
+            LOG(WARNING) << "File length changed before reading footer, path: "
+                         << file->path().native() << ", before: " << before_file_size
+                         << ", after: " << file_size;
+            bytes_read = std::min(file_size, INIT_META_SIZE);
+            RETURN_IF_ERROR(file->read_at(file_size - bytes_read, Slice(footer.data(), bytes_read),
+                                          &bytes_read, io_ctx));
+            if (!validate_magic_number(footer, bytes_read)) {
+                return Status::Corruption(
+                        "Invalid magic number in parquet file, bytes read: {}, file size: {}, "
```

Comment:
You can optimize here to not write this `Status::Corruption` twice.
For example, u can use a `Status` and return it at the end.